spec/compiler/parser/parser_spec.cr

```diff
@@ -1142,6 +1142,10 @@ describe "Parser" do
 
   it_parses "Foo.foo(count: 3).bar { }", Call.new(Call.new("Foo".path, "foo", named_args: [NamedArgument.new("count", 3.int32)]), "bar", block: Block.new)
 
+  it_parses "Foo?", Crystal::Generic.new(Path.global("Union"), ["Foo".path, Path.global("Nil")] of ASTNode)
+  it_parses "Foo::Bar?", Crystal::Generic.new(Path.global("Union"), [Path.new(%w(Foo Bar)), Path.global("Nil")] of ASTNode)
+  it_parses "Foo(T)?", Crystal::Generic.new(Path.global("Union"), [Generic.new("Foo".path, ["T".path] of ASTNode), Path.global("Nil")] of ASTNode)
```

Comment:
Unless `T` is a union itself since it is expanded as `R|S` and not `Union(R, S)`. Maybe union to_s should be always be expanded to `Union(...)` syntax, unless a flag in the `#to_s` is given, like the one that skips parens...